code/datums/chatmessage.dm

```diff
@@ -0,0 +1,342 @@
+	scheduled_destruction = world.time + (lifespan - CHAT_MESSAGE_EOL_FADE)
+	enter_subsystem()
+
+/**
+  * Applies final animations to overlay CHAT_MESSAGE_EOL_FADE deciseconds prior to message deletion,
+  * sets time for scheduling deletion and re-enters the runechat SS for qdeling
+  *
+  * Arguments:
+  * * fadetime - The amount of time to animate the message's fadeout for
+  */
+/datum/chatmessage/proc/end_of_life(fadetime = CHAT_MESSAGE_EOL_FADE)
+	eol_complete = scheduled_destruction + fadetime
+	animate(message, alpha = 0, time = fadetime, flags = ANIMATION_PARALLEL)
+	enter_subsystem(eol_complete) // re-enter the runechat SS with the EOL completion time to QDEL self
+
+/**
+  * Creates a message overlay at a defined location for a given speaker
+  *
+  * Arguments:
+  * * speaker - The atom who is saying this message
+  * * message_language - The language that the message is said in
+  * * raw_message - The text content of the message
+  * * spans - Additional classes to be added to the message
+  */
+/mob/proc/create_chat_message(atom/movable/speaker, datum/language/message_language, raw_message, list/spans, runechat_flags = NONE)
+	// Ensure the list we are using, if present, is a copy so we don't modify the list provided to us
+	spans = spans ? spans.Copy() : list()
+
+	// Check for virtual speakers (aka hearing a message through a radio)
+	var/atom/movable/originalSpeaker = speaker
+	if (istype(speaker, /atom/movable/virtualspeaker))
+		var/atom/movable/virtualspeaker/v = speaker
+		speaker = v.source
+		spans |= "virtual-speaker"
+
+	// Ignore virtual speaker (most often radio messages) from ourself
+	if (originalSpeaker != src && speaker == src)
+		return
+
+	// Display visual above source
+	if(runechat_flags & EMOTE_MESSAGE)
+		new /datum/chatmessage(raw_message, speaker, src, message_language, list("emote", "italics"))
+	else
+		new /datum/chatmessage(lang_treat(speaker, message_language, raw_message, spans, null, TRUE), speaker, src, message_language, spans)
+
+
+// Tweak these defines to change the available color ranges
+#define CM_COLOR_SAT_MIN	0.6
+#define CM_COLOR_SAT_MAX	0.7
+#define CM_COLOR_LUM_MIN	0.65
+#define CM_COLOR_LUM_MAX	0.75
+
+/**
+  * Gets a color for a name, will return the same color for a given string consistently within a round.atom
+  *
+  * Note that this proc aims to produce pastel-ish colors using the HSL colorspace. These seem to be favorable for displaying on the map.
+  *
+  * Arguments:
+  * * name - The name to generate a color for
+  * * sat_shift - A value between 0 and 1 that will be multiplied against the saturation
+  * * lum_shift - A value between 0 and 1 that will be multiplied against the luminescence
+  */
+/datum/chatmessage/proc/colorize_string(name, sat_shift = 1, lum_shift = 1)
+	// seed to help randomness
+	var/static/rseed = rand(1,26)
+
+	// get hsl using the selected 6 characters of the md5 hash
+	var/hash = copytext(md5(name + GLOB.round_id), rseed, rseed + 6)
```

Comment:
md5 is very slow and the thing that absolutely destroyed the server when the stat panel had it.

Reply:
This proc is only called if mob has no predefined color or it has no color and its name has changed.
I don't think it's gonna lag the server, especially if being called that rarely. 
How did you solve that in stat panel?

Reply:
Just didnt use it, the hash was just the name of the overlays themselves.